### DISCUSSION TITLE
index error get_untiled_crop with a tifffile input

### DISCUSSION THREAD

**Discussion by ViriatoII** [UTokyo-FieldPhenomics-Lab/EasyIDP#16] [Q&A] [answered]
I get an index error when using a normal tifffile input.

```---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
---> 26 _get_untiled_crop(page, 300, 200, 10, 10)

/tmp/ipykernel_460/304543409.py in _get_untiled_crop(page, i0, j0, h, w)
     12 
     13         for index in range(i0, i1):
---> 14             offset = page.dataoffsets[index]
     15             bytecount = page.databytecounts[index]
     16 

IndexError: tuple index out of range
```
The index generating error is a 300, my desired i0 coordinate. The .dataoffsets instance does not correspond to what you expect in an input page

My page input is simply read from a tifffile handle. It seems to work 
```
>page = tifffile.TiffFile(test_slide).pages[0]
>page.imagewidth, page.imagelength
(2142, 2134)
```

My page.dataoffsets only has 34 elements, it goes like this:
(8, 274184,  548360, ....)

So what's causing this difference and how can I solve it?

Cheers,
Ricardo

**Comment by HowcanoeWang** [comment, marked as answer]
Dear Ricardo,

Thanks for using our package, we have understood that you meet the open tifffile problem?

Do you mind if you would like to share the code that meets the error `26 _get_untiled_crop(page, 300, 200, 10, 10)`?

**Reply by ViriatoII** [reply]
Hey HawcanoeWang,

That was fast, thanks! 

```
def _get_untiled_crop(page, i0, j0, h, w):

        im_width = page.imagewidth
        im_height = page.imagelength

        i1, j1 = i0 + h, j0 + w
        if i0 < 0 or j0 < 0 or i1 >= im_height or j1 >= im_width:
            raise ValueError(f"Requested crop area is out of image bounds.{i0}_{i1}_{im_height}, {j0}_{j1}_{im_width}")
        
        out = np.empty((page.imagedepth, h, w, page.samplesperpixel), dtype=page.dtype)
        fh = page.parent.filehandle

        for index in range(i0, i1):
            offset = page.dataoffsets[index]
            bytecount = page.databytecounts[index]

            fh.seek(offset)
            data = fh.read(bytecount)

            tile, indices, shape = page.decode(data, index)

            out[:,index-i0,:,:] = tile[:,:,j0:j1,:]

        return out[0,:,:,:]
    
test_slide =  "33042_W0A1_P1X_R2_RO-Channel1_27052022-21-33-59_s1.ome.tiff"
page = tifffile.TiffFile(test_slide).pages[0]
_get_untiled_crop(page, 300, 200, 10, 10) 
```

Would you like the test_slide file as well? 
It's a 280Mb tiff file. Reading it with tifffile.imread generates a correct array with shape (32, 2134, 2142)   (32 is the z layers). I noticed tifffile.TiffFile() generated 32 pages, so I assume each page is a z layer.

**Reply by HowcanoeWang** [reply]
could you please check `page.is_tiled`? If it is tiled, i guess you need to use `_get_tiled_crop` instead.

BTW, it is not recommended to call hidden functions start with '_' directly. Please use `easyidp.geotiff.tifffile_crop(page, 300, 200, 10, 10)` instead. (please use the latest code by git clone, rather than `pip install easyidp`)

**Reply by HowcanoeWang** [reply]
> could you please check `page.is_tiled`? If it is tiled, i guess you need to use `_get_tiled_crop` instead.
> 
> BTW, it is not recommended to call hidden functions start with '_' directly. Please use `easyidp.geotiff.tifffile_crop(page, 300, 200, 10, 10)` instead. (please use the latest code by git clone, rather than `pip install easyidp`)

your `_get_untiled_crop` is code from v1.0, we have fixed some bug in it and please update to v2.0

**Reply by ViriatoII** [reply]
Oh, yes, _get_untiled_crop is defined twice in geotiff.py. I was copying the first one but the second one works!

(and the page was not tiled).

Thank you very much for the attention! 
Cheers,
Ricardo

**Reply by HowcanoeWang** [reply]
Glad that works for you! 
PS: `_get_untiled_crop` is not defined twice! They are `_get_tiled_crop` and `_get_untiled_crop`, dealing with `page.is_tiled==True` and `page.is_tiled==False` respectively.
